content/blog/2019/02/2019-02-22-serverless-jenkins.adoc

```diff
@@ -0,0 +1,126 @@
+“Serverless” Jenkins in the sense that we don’t have to provision the server for Jenkins to run.
+
+WARNING: The setup described in this post is highly experimental. I encourage you to try this out in a demo project, but be very cautions until further notice to use this on productive code. In this proof of concept, I’ll use a public GitHub repository and the free open-source offering by TravisCI. This setup is not suitable for commercial software.
+
+The pets vs cattle metaphor describes how approaches in managing servers differ.
+While you care for pets and treat them when they are unwell, cattle can be easily replaced.
+Your traditional Jenkins server is a pet because it is often configured manually, and replacing it is a major effort.
+For more background on this metaphor, http://cloudscaling.com/blog/cloud-computing/the-history-of-pets-vs-cattle/[click here].
+
+Before we’re getting into the technical details, let’s discuss why we would want to try this out in the first place.
+Running Jenkins on arbitrary CI/CD services, such as TravisCI seems very odd on first sight.
+On such services you’ll usually invoke your build tools like Maven or npm in a small script, and that will do your build.
+But in the enterprise world, both inside SAP and in the community, Jenkins has a huge market share.
+There are many shared libraries for Jenkins, providing pre-made build steps which would be expensive to re-implement for other platforms.
+Additionally, SAP S/4HANA Cloud SDK Pipeline is a ready to use pipeline based on Jenkins where you as the developer of an SAP S/4HANA extension application do not need to write and maintain the pipeline yourself.
+This means reduced costs and effort for you, while the quality of your application improves, for example due to the many https://github.com/SAP/cloud-s4-sdk-pipeline/blob/master/doc/pipeline/cloud-qualities.md[cloud qualities] which are checked out of the box.
+
+image:/images/post-images/2019-02-22/green.png
+
+
+Let me show you an experiment to see if we can get the best of both worlds. The goal is to get all the quality checks and the continuous delivery that the SAP S/4HANA Cloud SDK Pipeline provides us, without the need for a pet Jenkins server.
+
+How do we do that? The Jenkins project has a project called https://github.com/jenkinsci/jenkinsfile-runner[Jenkinsfile runner]. It is a command line tool that basically boots up a stripped-down Jenkins instance, creates and runs a single job, and throws away that instance once the job is done. As you might guess, there is some overhead in that process. This will add about 20 seconds to each build, which I found to be surprisingly fast, considering the usual startup time of a Jenkins server. For convenient consumption, we have packaged https://hub.docker.com/r/ppiper/jenkinsfile-runner[Jenkinsfile runner as a Docker image] which includes the Jenkins plugins that are required for SAP S/4HANA Cloud SDK Pipeline.
```

Comment:
Super long lines make it difficult to leave comments on specific parts. Note that single line breaks are ignored, so you can manually justify lines, or have one sentence per line.

Reply:
I'm a single line person, but I did not do it for the whole text which I imported from a wordpress post..